pyface/data_view/abstract_data_model.py

```diff
@@ -0,0 +1,272 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+""" Provides an AbstractDataModel ABC for Pyface data models.
+
+This module provides an ABC for all data view data models.  This specifies
+the API that the data view widgets expect, and which the underlying
+data is adapted to by the concrete implementations.  Data models are intended
+to be toolkit-independent, and be able to adapt any approximately tabular or
+nested data structure to what the data view system expects.
+"""
+from abc import abstractmethod
+
+from traits.api import ABCHasStrictTraits, Event, Instance
+
+from .index_manager import AbstractIndexManager
+
+
+class AbstractDataModel(ABCHasStrictTraits):
+    """ Abstract base class for Pyface data models.
+
+    The data model API is intended to provide a common API for heirarchical
+    and tabular data.  This class is concerned with the structure, type and
+    values provided by the data, but not with how the data is presented.
+
+    Row and column indices are represented by sequences (usually lists) of
+    integers, specifying the index at each level of the heirarchy.  The root
+    row and column are represented by empty lists.
+
```

Comment:
Row and column indices are sequences, not lists; we just happen to be using lists a lot in the implementation because they are convenient (building indices by appending to a list vs. adding new elements to extend a tuple).  I just did some performance testing, and found that:
```
result = ()
while ...:
    result += (next,)
```
is actually comparable in performance to:
```
result = []
while ...:
    result.append(next)
```
I had thought that the latter would have significantly better performance, but apparently not.  So I think we can start using tuples instead of lists in the implementation (I'd only used lists for performance).  I'm not sure that I want to _rely_ on the values supplied being tuples, however, at least not yet.

However, we are not going to get away from len(...) operations as that is essentially giving you how deep you are in the hierarchy.  If you look at the `ArrayDataModel` implementation, sometimes we are testing with len(...) == 0, but other times we are testing with len(...) < ... to determine if we are a leaf node or not.

Given this it is better to use len(...) consistently, rather than sometimes using len and sometimes testing for equality. (Ieva pointed this out earlier in the review process, and she is right).

I also don't think I want a custom object either; I think it will just add complexity without much advantage.